Azure helper: Increase Azure Endpoint HTTP retries

johnsonshi opened the pull request:
## Proposed Commit Message
Azure helper: Increase Azure Endpoint HTTP retries

Increase Azure Endpoint HTTP retries to handle
occasional platform network blips.

Introduce a common method `http_with_retries`
in the azure.py helper, which will serve as
the common HTTP request handler for
all HTTP requests with the Azure endpoint.
This method has builtin retries and
reporting diagnostics logic.

## Additional Context
Additional context above.

## Test Steps

#### Ensure no regressions in pre-existing Azure endpoint interaction for happy path code
* Create an Azure VM from the latest Ubuntu 18.04 LTS image.
* Install this PR's cloud-init.
* Capture the VM into a custom image.
* Deploy a VM from this custom image. Deployment succeeds.
* No warnings/errors. Manually inspected cloud-init logs and ensured interaction with the Azure endpoint was seamless.

#### Ensure that long delay log message is logged, if there is a long delay (>5 sec) that eventually succeeds during interaction with the Azure endpoint.
* Same steps as above. Modify `http_with_retries` to fail during the first 15 attempts, and only succeed at the 16th attempt.
* Capture the VM into a custom image.
* Deploy a VM from this custom image. Deployment succeeds.
* No warnings/errors. Manually inspected cloud-init logs and ensured interaction with the Azure endpoint succeeded but with the new success diagnostic message present.

#### Ensure that error messages are logged upon failure despite max attempt retries
* Same as above, except modify `http_with_retries` so that it fails after max attempts.
* Inspect cloud-init logs to ensure that the new failure diagnostic message is present.

## Checklist:
 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/hacking.html)
 - [x] I have updated or added any unit tests accordingly
 - [x] I have updated or added any documentation accordingly

johnsonshi commented:
Pending internal discussions on what the error messages should be.

mitechie commented:
@johnsonshi is this ready for review? Are we happy with the error messages and such at this point? 

Thanks!

johnsonshi commented:
@mitechie yes we've discussed internally and these will be the final error messages.

blackboxsw commented on cloudinit/sources/helpers/azure.py:
```diff
@@ -258,6 +258,51 @@ def _get_dhcp_endpoint_option_name():
+        LOG.warning(
+            'infinite kwarg passed in for communication with Azure endpoint.')
+
+    for attempt in range(max_readurl_attempts):
```
We could change this call to range call to be 1-based instead of 0-based, then you don't have to add 1 to the attempt value below
```suggestion
    for attempt in range(1, max_readurl_attempts + 1):
```

blackboxsw commented on cloudinit/sources/helpers/azure.py:
```diff
@@ -258,6 +258,51 @@ def _get_dhcp_endpoint_option_name():
+
+            report_diagnostic_event(
+                'Successful HTTP request with Azure endpoint %s after '
+                '%d attempts' % (url, attempt + 1),
```
```suggestion
                '%d attempts' % (url, attempt),
```

johnsonshi replied:
Done. Same commit as above: https://github.com/canonical/cloud-init/pull/619/commits/81b3a9e18f095549ffe61ed7eba1c8987d048b16

blackboxsw commented on cloudinit/sources/helpers/azure.py:
```diff
@@ -258,6 +258,51 @@ def _get_dhcp_endpoint_option_name():
+
+        except Exception as e:
+            exc = e
+            if (attempt + 1) % periodic_logging_attempts == 0:
```
```suggestion
            if (attempt) % periodic_logging_attempts == 0:
```

blackboxsw commented on cloudinit/sources/helpers/azure.py:
```diff
@@ -258,6 +258,51 @@ def _get_dhcp_endpoint_option_name():
+                report_diagnostic_event(
+                    'Failed HTTP request with Azure endpoint %s during '
+                    'attempt %d with exception: %s' %
+                    (url, attempt + 1, e),
```
```suggestion
                    (url, attempt, e),
```

johnsonshi replied:
Done. Same commit as above: https://github.com/canonical/cloud-init/pull/619/commits/81b3a9e18f095549ffe61ed7eba1c8987d048b16

blackboxsw commented on cloudinit/sources/helpers/azure.py:
```diff
@@ -258,6 +258,51 @@ def _get_dhcp_endpoint_option_name():
 
+@azure_ds_telemetry_reporter
+def http_with_retries(url, **kwargs) -> str:
+    exc = None
```
Might be worth a small docstring describing the intent here being periodic custom telemetry logging which readurl doesn't provide.

johnsonshi replied:
Done. https://github.com/canonical/cloud-init/pull/619/commits/8dfe0598fc65e1aae13fb6641f7cba7eae918a70

blackboxsw commented on cloudinit/sources/helpers/azure.py:
```diff
@@ -258,6 +258,51 @@ def _get_dhcp_endpoint_option_name():
+
+    # remove kwargs that cause url_helper.readurl to retry,
+    # since we are already implementing our own retry logic.
+    if 'retries' in kwargs:
+        kwargs.pop('retries', None)
+        LOG.warning(
+            'retries kwarg passed in for communication with Azure endpoint.')
```
Since we aren't using the value of retries, no need to if and pop. Let's also just mention in logs what we are doing with that unexpected arg. "Ignoring..."

```suggestion
    if kwargs.pop('retries', None):
        LOG.warning(
            'Ignoring retries kwarg passed in for communication with Azure endpoint.')
```

johnsonshi replied:
Done. https://github.com/canonical/cloud-init/pull/619/commits/d8211a3cf6a514996ec6e5f3f45f6862ef52960a

blackboxsw commented on cloudinit/sources/helpers/azure.py:
```diff
@@ -258,6 +258,51 @@ def _get_dhcp_endpoint_option_name():
+        kwargs.pop('retries', None)
+        LOG.warning(
+            'retries kwarg passed in for communication with Azure endpoint.')
+    if 'infinite' in kwargs:
+        kwargs.pop('infinite', None)
+        LOG.warning(
+            'infinite kwarg passed in for communication with Azure endpoint.')
```
Same as above
```suggestion
    if kwargs.pop('infinite', None):
        LOG.warning(
            'Ignoring infinite kwarg passed in for communication with Azure endpoint.')
```

blackboxsw reviewed (changes requested):
Thank you so much @johnsonshi for this branch. Minor nits on review. If this is acceptable, I can apply them or you can then we will land it.

johnsonshi commented:
Thanks you so much @blackboxsw! I've addressed the requested changes above. Let's wait for https://github.com/canonical/cloud-init/pull/594 to land before merging this.

https://github.com/canonical/cloud-init/pull/594 is also about to be merged. Just addressed OddBloke's final set of minor requested changes and NITs in that other PR, so it should be mergeable once he reviews it.

Given that https://github.com/canonical/cloud-init/pull/594 is more complicated than this PR, it will be easier to fix this PR if there are any merge conflicts.

blackboxsw reviewed (approved):
Thanks @johnsonshi this it good to land today.
